lib/gossip/damper.js

```diff
@@ -0,0 +1,542 @@
+        });
+        return false;
+    }
+
+    if (this.flappers[flapper.address]) {
+        this.logger.debug('ringpop damper already added flapper', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    // TODO Limit number of flappers?
+    // TODO Expire flappers?
+
+    this.flappers[flapper.address] = flapper;
+    this.logger.debug('ringpop damper added flapper', {
+        local: this.ringpop.whoami(),
+        flapper: flapper.address
+    });
+    this.ringpop.stat('increment', 'damper.flapper.added');
+    this.ringpop.stat('gauge', 'damper.flappers',
+        Object.keys(this.flappers.length));
+
+    if (this._getFlapperCount() === 1) {
+        this._startDampTimer();
+    }
+
+    return true;
+};
+
+Damper.prototype.dampMember = function dampMember(addr)  {
+    var self = this;
+    var member = this.ringpop.membership.findMemberByAddress(addr);
+
+    if (!member) {
+        this.logger.warn('ringpop damper cannot damp member; member does not exist', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    if (!this.hasFlapper(member)) {
+        this.logger.warn('ringpop damper cannot damp member; member is not flappy', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    this.ringpop.membership.makeDamped(member.address);
+
+    // Don't start anymore damping subprotocols for a member
+    // that is already damped.
+    this.removeFlapper(member.address);
+    this.addDampedMember(member.address);
+
+    if (!this.expirationTimer) {
+        scheduleExpiration();
+        this.logger.info('ringpop started damped member expiration timer', {
+            local: this.ringpop.whoami()
+        });
+    }
+
+    return true;
+
+    function scheduleExpiration() {
+        self.expirationTimer =
+            self.timers.setTimeout(function onTimeout() {
+                self.expireDampedMembers();
+                scheduleExpiration();
+        }, self.ringpop.config.get('dampedMemberExpirationInterval'));
+    }
+};
+
+Damper.prototype.expireDampedMembers = function expireDampedMembers() {
+    var undampedMembers = [];
+    var suppressDuration = this.ringpop.config.get('dampScoringSuppressDuration');
+    var keys = Object.keys(this.dampedMembers);
+    for (var i = 0; i < keys.length; i++) {
+        var key = keys[i];
+        var memberDamping = this.dampedMembers[keys[i]];
+        var dampedDuration = this.Date.now() - memberDamping.timestamp;
+        if (dampedDuration >= suppressDuration) {
+            delete this.dampedMembers[key];
+            undampedMembers.push({
+                member: key,
+                timeDamped: dampedDuration
+            });
+        }
+    }
+
+    if (undampedMembers.length === 0) {
+        return undampedMembers;
+    }
+
+    this.emit('event', new events.DampedMemberExpirationEvent(undampedMembers));
```

Comment:
`event` is not the best name here in my opinion. Can we be more specific what kind of event this is, eg `expired` ?